Contextual fields combine for SDKs (#743)
rudderlabs/rudderstack-docs: docs/event-spec/standard-events/common-fields.mdx

## Contextual fields

> The contextual fields give additional useful context about a particular event. The following table lists all the contextual fields with their detailed description:

Reviewer: ```suggestion
The contextual fields give additional context about a particular event. The following table lists all the contextual fields with their detailed description:
```

| `app.version`      | -   | Yes | Yes | Version of the application. | 
| `app.build`        | -   | Yes | Yes | Build of the application. |
| `campaign.name`    | Yes | -   | -   | Name of the campaign.  |
| `campaign.source`  | Yes | -   | -   | Source of the campaign. |
| `campaign.medium`  | Yes | -   | -   | Medium of the campaign. |
| `campaign.term`    | Yes | -   | -   | Term of the campaign. |
| `campaign.content` | Yes | -   | -   | Content of the campaign. |
| `device.type`      | -   | Yes | Yes   | Type of the device. |
| `device.id`        | -   | Yes | Yes | ID of the device. |
| `device.advertisingId` | - | Yes | Yes | Advertising ID of the device. |
| `device.adTrackingEnabled` | - | - | Yes | If ad tracking is enabled on the device.   |
| `device.manufacturer` | - | Yes | Yes | Manufacturer of the device.|
| `device.model` | - | Yes | Yes | Model of the device. |
| `device.name` | - | Yes | Yes | Name of the device. |
| `library.name` | Yes | Yes | Yes | Name of the library. |
| `library.version` | Yes | Yes | Yes | Version of the library. |
| `ip` | - | Yes | - |  IP address of the user. |
| `locale` | Yes | Yes | Yes | Locale string of the user. |

> | `network.bluetooth` | - | Yes (if permission is granted) | No | Bluetooth information about the network connection. |

Reviewer 2: `Bluetooth information about the network connection.` - @pallabmaiti, does this make sense? "about the network connection"

Reviewer 3: Right @saikumarrs it doesn't make sense. @ishachopra0401 for iOS, it will be `-`

Reviewer 2: I meant "about the network connection". How's Bluetooth connected to the network?

Author: @saikumarrs: I had a discussion with @pallabmaiti and have modified the description to "Bluetooth information" only.

| `network.carrier` | - | Yes | Yes | Carrier information about the network connection. |
| `network.cellular` | - | Yes | Yes | Cellular information about the network connection. |

> | `network.wifi` | - | Yes | Yes | Wifi information about the network connection. |

Reviewer 2: ```suggestion
| `network.wifi` | - | Yes | Yes | WiFi information about the network connection. |
```

| `page.initial_referring_domain` | Yes | - |  - | Initial referring domain of the current page in the browser. |
| `page.referrer` | Yes | - |  - | Referrer of the current page in the browser. |
| `page.referring_domain` | Yes | - |  - | Referring domain of the current page in the browser. |
| `page.search` | Yes | - |  - | Search of the current page in the browser.
| `page.title` | Yes | - |  - | Title of the current page in the browser. |
| `page.url` | Yes | - |  - | URL of the current page in the browser. |
| `screen.density` | Yes | Yes | - | Density of the device's screen. |
| `screen.height` | Yes | Yes | Yes | Height of the device's screen. |
| `screen.width` | Yes | Yes | Yes | Width of the device's screen. |
| `screen.innerWidth` | Yes |   -   |   -   | Inner width of the device's screen. |
| `screen.innerHeight` | Yes |   -   |   -   | Inner height of the device's screen. |
| `traits` | - | Yes | Yes | Traits of the user. |
| `userAgent` | Yes | Yes | - | User agent of the device making the request. |
| `timezone` | - | Yes | Yes | Information about the user's timezone. |

<div class="infoBlock">

> For Android SDK V1.6.0 and above, the bluetooth status is collected in the <code class="inline-code">network.bluetooth</code> field only if the bluetooth permission is present in the app.
> For iOS SDK V1.6.3 and above, the bluetooth status is not collected in the <code class="inline-code">network.bluetooth</code> field as it’s a run time permission.

Reviewer 2: ```suggestion
For Android SDK V1.6.0 and above, the Bluetooth status is collected in the <code class="inline-code">network.bluetooth</code> field only if the Bluetooth permission is present in the app.
For iOS SDK V1.6.3 and above, the Bluetooth status is not collected in the <code class="inline-code">network.bluetooth</code> field as it’s a run time permission.
```

</div>

The following sample payload highlights the usage of the common and contextual fields in web and mobile modes:

<Tabs>
  <TabList>
    <Tab>JavaScript</Tab>
    <Tab>Android</Tab>
    <Tab>iOS</Tab>
  </TabList>
    <TabPanels>
      <TabPanel>
<span>

```json
{

>   "anonymousId": "7e22166a4dab652f",

Reviewer 2: The example anonymous ID does not look real. Please fix other examples too in this page.

Author: @saikumarrs Can you please suggest a sample one?

  "channel": "web",
  "context": {
    "campaign": {
      "name": "sales campaign",
      "source": "google",
      "medium": "medium",
      "term": "event data",
      "content": "Make sense of the modern data stack"
    },
    "library": {

>       "name": "com.rudderstack.javascript.sdk.core",

Reviewer 2: This is incorrect.

Author: @saikumarrs Please suggest what should go here instead

Reviewer 2: `RudderLabs JavaScript SDK`

>       "version": "1.0.0"

Reviewer 2: Use `2.9.1` as it's the latest version.

    },
    "locale": "en-US",
    "page": {
      "path": "/best-seller/1",
      "initial_referrer": "https://www.google.com/search",
      "initial_referring_domain": "google.com",
      "referrer": "https://www.google.com/search?q=estore+bestseller",
      "referring_domain": "google.com",
      "search": "estore bestseller",
      "title": "The best sellers offered by EStore",
      "url": "https://www.estore.com/best-seller/1"
    },

>     "screen": { "density": 420, "height": 1794, "width": 1080, "innerHeight": 200, "innerWidth": 100 },

Reviewer 2: Formatting issue.

  },
  "event": "Product Reviewed",
  "integrations": { "All": true },
  "messageId": "1578564113557-af022c68-429e-4af4-b99b-2b9174056383",
  "properties": {
    "review_id": "review_id_1",
    "product_id": "product_id_1",
    "rating": 2.0,
    "review_body": "Sample Review Body"
  },
  "originalTimestamp": "2020-01-09T10:01:53.558Z",
  "type": "track",
  "sentAt": "2020-01-09T10:02:03.257Z"
}
```
</span>
      </TabPanel>
      <TabPanel>
<span>

> ```json

Reviewer 2: I'd add two different example payloads for Android and iOS SDKs.

Author: done

| `campaign` | Object | Gives detailed information about campaigns, like `name`, `source`, `medium`, `content` and `term`. |

<div class="warningBlock">

> <strong>All of the above contextual fields are not automatically collected by the RudderStack SDKs.</strong> Refer to the following section for more information on which SDK automatically populates what contextual fields.

Reviewer: ```suggestion
<strong>All of the above contextual fields are not automatically collected by the RudderStack SDKs.</strong> Refer to the following section for more information on the contextual fields automatically populated by the RudderStack SDKs.
```

| `page.referrer` | Yes | - |  - | Referrer of the current page in the browser. |
| `page.referring_domain` | Yes | - |  - | Referring domain of the current page in the browser. |
| `page.search` | Yes | - |  - | Search of the current page in the browser.
| `page.title` | Yes | - |  - | Title of the current page in the browser. |
| `page.url` | Yes | - |  - | URL of the current page in the browser. |
| `screen.density` | Yes | Yes | - | Density of the device's screen. |
| `screen.height` | Yes | Yes | Yes | Height of the device's screen. |
| `screen.width` | Yes | Yes | Yes | Width of the device's screen. |
| `screen.innerWidth` | Yes |   -   |   -   | Inner width of the device's screen. |
| `screen.innerHeight` | Yes |   -   |   -   | Inner height of the device's screen. |
| `traits` | - | Yes | Yes | Traits of the user. |
| `userAgent` | Yes | Yes | - | User agent of the device making the request. |
| `timezone` | - | Yes | Yes | Information about the user's timezone. |

<div class="infoBlock">
For Android SDK V1.6.0 and above, the Bluetooth status is collected in the <code class="inline-code">network.bluetooth</code> field only if the Bluetooth permission is present in the app.

> For iOS SDK V1.6.3 and above, the Bluetooth status is not collected in the <code class="inline-code">network.bluetooth</code> field as it’s a run time permission.

Reviewer: ```suggestion
For iOS SDK V1.6.3 and above, the Bluetooth status is not collected in the <code class="inline-code">network.bluetooth</code> field as it’s a run-time permission.
```

      "version": "2.9.1"
    },
    "locale": "en-US",
    "page": {
      "path": "/best-seller/1",
      "initial_referrer": "https://www.google.com/search",
      "initial_referring_domain": "google.com",
      "referrer": "https://www.google.com/search?q=estore+bestseller",
      "referring_domain": "google.com",
      "search": "estore bestseller",
      "title": "The best sellers offered by EStore",
      "url": "https://www.estore.com/best-seller/1"
    },
    "screen": { 
      "density": 420, 
      "height": 1794, 
      "width": 1080, 
      "innerHeight": 200, 
      "innerWidth": 100 
      },
    "userAgent": "Dalvik/2.1.0 (Linux; U; Android 9; Android SDK built for x86 Build/PSR1.180720.075)"
  },
  "event": "Product Reviewed",
  "integrations": { 
    "All": true 
    },
  "messageId": "1578564113557-af022c68-429e-4af4-b99b-2b9174056383",
  "properties": {
    "review_id": "review_id_1",
    "product_id": "product_id_1",
    "rating": 2.0,
    "review_body": "Sample Review Body"
  },
  "originalTimestamp": "2020-01-09T10:01:53.558Z",
  "type": "track",
  "sentAt": "2020-01-09T10:02:03.257Z"
}
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```json

> {

Reviewer 3: Here is a sample JSON:
`
{
        "messageId": "1643629072-8ce81faf-7489-4508-b21b-659e81510991",
        "anonymousId": "anonymous_id",
        "userId": "new_user_id",
        "channel": "mobile",
        "event": "Application Installed",
        "context": {
            "screen": {
                "density": 3,
                "width": 844,
                "height": 390
            },
            "os": {
                "name": "iOS",
                "version": "15.2"
            },
            "locale": "en-US",
            "app": {
                "version": "1.0",
                "namespace": "com.rudderstack.ios.test.objc",
                "name": "RudderSampleAppObjC",
                "build": "1"
            },
            "device": {
                "attTrackingStatus": 0,
                "manufacturer": "Apple",
                "id": "42f0686d-564d-4e2d-815c-a9380dd8b70f",
                "model": "iPhone",
                "adTrackingEnabled": true,
                "type": "iOS",
                "token": "your_device_token",
                "advertisingId": "advertisement_Id",
                "name": "iPhone 12"
            },
            "traits": {
                "id": "new_user_id",
                "anonymousId": "anonymous_id",
                "userId": "new_user_id",
                "firstname": "First Name"
            },
            "library": {
                "name": "rudder-ios-library",
                "version": "1.5.0"
            },
            "timezone": "Asia\/Kolkata",
            "network": {
                "bluetooth": false,
                "wifi": true,
                "carrier": "unavailable",
                "cellular": false
            },
            "externalId": [{
                "type": "brazeExternalId",
                "id": "some_external_id_1"
            }]
        },
        "originalTimestamp": "2022-01-31T11:37:52.481Z",
        "properties": {
            "version": "1.0"
        },
        "type": "track",
        "integrations": {
            "All": true
        },
        "sentAt": "2022-01-31T11:38:08.731Z"
    }
`

      "density": 420, 
      "height": 1794, 
      "width": 1080 
      },
     "traits": { 
      "anonymousId": "7e32188a4dab669f" 
      },
    "timezone": "Asia/Mumbai",
    "userAgent": "Dalvik/2.1.0 (Linux; U; Android 9; Android SDK built for x86 Build/PSR1.180720.075)"
  },
  "event": "Product Reviewed",
  "integrations": { 
    "All": true 
    },
  "messageId": "1578564113557-af022c68-429e-4af4-b99b-2b9174056383",
  "properties": {
    "review_id": "review_id_1",
    "product_id": "product_id_1",
    "rating": 2.0,
    "review_body": "Sample Review Body"
  },
  "originalTimestamp": "2020-01-09T10:01:53.558Z",
  "type": "track",
  "sentAt": "2020-01-09T10:02:03.257Z"
}
```
</span>
      </TabPanel>
      <TabPanel>
<span>

```json
{
  "anonymousId": "7e32188a4dab669f",
  "channel": "mobile",
  "context": {
    "app": {
      "name": "RudderiOSClient",
      "version": "1.0",
      "build": "1"
    },
    "device": {
      "type": "iOS",
      "id": "7e32188a4dab669f",
      "advertisingId":"435o4GRlm",
      "adTrackingEnabled": false,
      "manufacturer": "Google",
      "model": "iOS SDK built for x86",
      "name": "generic_x86",
    },
    "library": {

>       "name": "com.rudderstack.iOS.sdk.core",

Reviewer 3: For iOS, `rudder-ios-library`

The following section highlights all the contextual fields that are **automatically collected** and populated by each of the RudderStack SDKs:

| Field | JavaScript | Android | iOS |
| :--- | :--- |:--- |:--- |
| `app.name`         | -   | Yes | Yes |
| `app.version`      | -   | Yes | Yes |
| `app.build`        | -   | Yes | Yes |
| `campaign.name`    | Yes | -   | -   |
| `campaign.source`  | Yes | -   | -   |
| `campaign.medium`  | Yes | -   | -   |
| `campaign.term`    | Yes | -   | -   |
| `campaign.content` | Yes | -   | -   |
| `device.type`      | -   | Yes | Yes   |
| `device.id`        | -   | Yes | Yes |
| `device.advertisingId` | - | Yes | Yes |
| `device.adTrackingEnabled` | - | - | Yes |
| `device.manufacturer` | - | Yes | Yes |
| `device.model` | - | Yes | Yes |
| `device.name` | - | Yes | Yes |
| `library.name` | Yes | Yes | Yes |
| `library.version` | Yes | Yes | Yes |
| `ip` | - | Yes | - |
| `locale` | Yes | Yes | Yes |
| `network.bluetooth` | - | Yes | Yes | 
| `network.carrier` | - | Yes | Yes |
| `network.cellular` | - | Yes | Yes |
| `network.wifi` | - | Yes | Yes |
| `os.name` | - | Yes | Yes |
| `os.version` | - | Yes | Yes |
| `page.path` | Yes | - |  - |

> | `page.referrer` | Yes | - |  - |

Reviewer 4: ```suggestion
| `page.initial_referrer` | Yes | - |  - |
| `page.initial_referring_domain` | Yes | - |  - |
| `page.referrer` | Yes | - |  - |
| `page.referring_domain` | Yes | - |  - |
```

| `page.search` | Yes | - |  - |
| `page.title` | Yes | - |  - |
| `page.url` | Yes | - |  - |
| `screen.density` | Yes | Yes | - |
| `screen.height` | Yes | Yes | Yes |

> | `screen.width` | Yes | Yes | Yes |

Reviewer 4: ```suggestion
| `screen.width` | Yes | Yes | Yes |
| `screen.innerWidth` | Yes |   -   |   -   |
| `screen.innerHeight` | Yes |   -   |   -   |
```

| `campaign` | Object | Gives detailed information about campaigns, like `name`, `source`, `medium`, `content` and `term`. |

<div class="warningBlock">
<strong>All of the above contextual fields are not automatically collected by the RudderStack SDKs.</strong> Refer to the following section for more information on which SDK automatically populates what contextual fields.
</div>

The following section highlights all the contextual fields that are **automatically collected** and populated by each of the RudderStack SDKs:

| Field | JavaScript | Android | iOS |
| :--- | :--- |:--- |:--- |
| `app.name`         | -   | Yes | Yes |
| `app.version`      | -   | Yes | Yes |

> | `app.build`        | -   | Yes | Yes |

Reviewer 2: @MoumitaM This is not true. We only set default values for those fields, and they are not the actual application information.

> | `app.build`        | -   | Yes | Yes |

Reviewer 4: Yes, since we fill up these values I thought of adding it. @ishachopra0401 ignore this.

| `campaign.name`    | Yes | -   | -   |
| `campaign.source`  | Yes | -   | -   |
| `campaign.medium`  | Yes | -   | -   |
| `campaign.term`    | Yes | -   | -   |
| `campaign.content` | Yes | -   | -   |

> | `device.type`      | -   | Yes | -   |

Reviewer 3: ```suggestion
| `device.type`      | -   | Yes | Yes   |
```

| `device.id`        | -   | Yes | Yes |
| `device.advertisingId` | - | Yes | Yes |
| `device.adTrackingEnabled` | - | - | Yes |
| `device.manufacturer` | - | Yes | Yes |
| `device.model` | - | Yes | Yes |
| `device.name` | - | Yes | - |
| `library.name` | Yes | Yes | Yes |
| `library.version` | Yes | Yes | Yes |

> | `ip` | - | Yes | Yes |

Reviewer 3: ```suggestion
| `ip` | - | Yes | - |
```

| `locale` | Yes | Yes | Yes |
| `location.latitude` | - | - | - |
| `location.longitude` | - | - |  - |
| `location.speed` | - | - | - |

> | `network.bluetooth` | - | Yes | - | 

Reviewer 3: ```suggestion
| `network.bluetooth` | - | Yes | Yes | 
```

| `network.carrier` | - | Yes | Yes |
| `network.cellular` | - | Yes | Yes |
| `network.wifi` | - | Yes | Yes |
| `os.name` | - | Yes | Yes |
| `os.version` | - | Yes | Yes |
| `page.path` | Yes | - |  - |
| `page.referrer` | Yes | - |  - |
| `page.search` | Yes | - |  - |
| `page.title` | Yes | - |  - |
| `page.url` | Yes | - |  - |

> | `screen.density` | - | Yes | - |

Reviewer 2: Adding to that, we have `innerHeight` and `innerWidth` too. @MoumitaM, check if there are any such fields in the SDK.

> | `screen.density` | - | Yes | - |

Author: @MoumitaM Can you please provide details related to these additional fields?

> | `screen.density` | - | Yes | - |

Reviewer 4: Check out the below suggestions.

</div>

The following section highlights all the contextual fields that are automatically collected and populated by each of the RudderStack SDKs:

| Field | JavaScript | Android | iOS |
| :--- | :--- |:--- |:--- |:--- |:--- |
| `app.name`         | -   | Yes | Yes |
| `app.version`      | -   | Yes | Yes |
| `app.build`        | -   | Yes | Yes |
| `campaign.name`    | Yes | -   | -   |
| `campaign.source`  | Yes | -   | -   |
| `campaign.medium`  | Yes | -   | -   |
| `campaign.term`    | Yes | -   | -   |
| `campaign.content` | Yes | -   | -   |
| `device.type`      | -   | Yes | -   |
| `device.id`        | -   | Yes | Yes |
| `device.advertisingId` | - | Yes | Yes |
| `device.adTrackingEnabled` | - | - | Yes |
| `device.manufacturer` | - | Yes | Yes |
| `device.model` | - | Yes | Yes |
| `device.name` | - | Yes | - |
| `library.name` | Yes | Yes | Yes |
| `library.version` | Yes | Yes | Yes |
| `ip` | - | Yes | Yes |
| `locale` | Yes | Yes | Yes |
| `location.latitude` | - | - | - |
| `location.longitude` | - | - |  - |

> | `location.speed` | - | - | - |

Author: @saikumarrs these 3 `location` fields are not collected for any of the SDKs, should we remove them?

Reviewer 2: Makes sense.
